pallets/nft-marketplace/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+			instance_id: T::InstanceId,
+			currency: CurrencyOf<T>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let seller = ensure_signed(origin.clone())?;
+
+			// Check that the seller is the owner of the nft
+			ensure!(
+				Self::is_owner(seller.clone(), class_id, instance_id)?,
+				Error::<T>::NotOwner,
+			);
+
+			// Check that the nft is not yet for sale
+			ensure!(
+				!<Gallery<T>>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyForSale
+			);
+
+			// Transfer the NFT to the parachain account
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &Self::account())
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Put the nft for sale
+			let sale = Sale {
+				seller,
+				price: AskingPrice { currency, amount },
+			};
+			<Gallery<T>>::insert(class_id, instance_id, sale.clone());
+			Self::deposit_event(Event::ForSale(sale));
+
+			Ok(())
+		}
+
+		/// Remove an NFT from sale
+		///
+		/// The sellers of an NFT that is for sale can call this extrinsic to reclaim
+		/// ownership over their NFT and remove it from sale.
+		///
+		/// Fails if
+		///   - the nft is not for sale
+		///   - `origin` is not the seller of the NFT
+		///   - transferring the ownership of the NFT back to the seller fails
+		#[pallet::weight(10_000_000)]
+		pub fn remove(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(who == sale.seller, Error::<T>::NotOwner);
+
+			// Transfer the NFT back to the seller, i.e., the original owner of this NFT
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &sale.seller)
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Remove the NFT from the gallery
+			<Gallery<T>>::remove(class_id, instance_id);
+			Self::deposit_event(Event::Removed {
+				class_id,
+				instance_id,
+			});
+
+			Ok(())
+		}
+
+		/// Buy the given nft
+		///
+		/// Fails if
+		///   - the NFT is not for sale
+		///   - `origin` is the seller of the NFT
+		///   - `origin` does not have enough balance of the currency the nft is being sold in
+		///   - transferring the asking price from the buyer to the seller fails
+		///   - transferring the nft to the buyer fails
+		#[pallet::weight(10_000_000)]
+		#[transactional]
+		pub fn buy(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let buyer = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(buyer != sale.seller, Error::<T>::IsSeller);
+
+			// Make sure the buyer can pay for the NFT
+			T::Fungibles::can_withdraw(sale.price.currency, &buyer, sale.price.amount)
```

Comment:
This check isn't technically necessary - the `transfer` should fail with a descriptive error. Once that detailed error is exposed instead of a generic error we don't need to do this.

I would pretty much only ever use `can_withdraw` in a non-transactional extrinsic where I want to check the balance early and skip a bunch of on-chain work that would be done before the final transfer, or something like that. Otherwise just let the transfer fail - the extra check costs more weight.

In addition, `InsufficientBalance` is pretty generic - it could be that the user has enough tokens but they're locked, for example. The underlying error of a failed transfer will include more details than we can provide by just querying if withdrawal is possible.

Reply:
Thank you for taking the time for educating me on this, I _really_ appreciate it 🙌